File: src/main/java/com/telegram/folobot/service/UserService.java
package com.telegram.folobot.service;

import com.telegram.folobot.FoloBot;
import com.telegram.folobot.dto.FoloPidorDto;
import com.telegram.folobot.dto.FoloUserDto;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.meta.api.methods.groupadministration.GetChatMember;
import org.telegram.telegrambots.meta.api.objects.User;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;

import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static com.telegram.folobot.Utils.printExeptionMessage;

@Component
@RequiredArgsConstructor
@Accessors(chain = true)
public class UserService {
    private final FoloUserService foloUserService;

    @Setter
    private FoloBot foloBot;

    /**
     * Получение имени пользователя
     *
     * @param user {@link User}
     * @return Имя пользователя
     */
    public String getUserName(User user) {
        if (!Objects.isNull(user)) {
            return Stream.of(Stream.of(user.getFirstName(), user.getLastName())
                                    .filter(Objects::nonNull)
                                    .collect(Collectors.joining(" ")),
                            user.getUserName())
                    .filter(Objects::nonNull)
                    .findFirst()
                    .orElse(null);
        }
        return null;
    }

    /**
     * Получение имени пользователя
     *
     * @param user {@link User}
     * @return Имя пользователя
     */
    public String getFoloUserName(User user) {
        FoloUserDto foloUser = foloUserService.findById(user.getId());
        // По тэгу
        String userName = foloUser.getTag();
        if (userName.isEmpty()) userName = getUserName(user);
        // По сохраненному имени
        if (userName == null || userName.isEmpty()) userName = foloUser.getName();
        // Если не удалось определить
        if (userName.isEmpty()) userName = "Загадочный незнакомец";
        return userName;
    }

    /**
     * Получение имени фолопидора
     *
     * @param foloPidorDto {@link FoloPidorDto}
     * @return Имя фолопидора
     */
    public String getFoloUserName(FoloPidorDto foloPidorDto) {
        // По тэгу
        String userName = foloPidorDto.getTag();
        // По пользователю
        if (userName.isEmpty()) userName = getUserName(getUserById(foloPidorDto.getId().getUserId()));
        // По сохраненному имени
        if (userName == null || userName.isEmpty()) userName = foloPidorDto.getName();
        // Если не удалось определить
        if (userName.isEmpty()) userName = "Загадочный незнакомец";
        return userName;
    }

    private User getUserById(Long userid) {
        try {
            return foloBot.execute(new GetChatMember(Long.toString(userid), userid)).getUser();
        } catch (TelegramApiException ignored) {
            return null;
        }
    }

    /**
     * Получение кликабельного имени пользователя
     *
     * @param user {@link User}
     * @return Имя пользователя
     */
    private String getFoloUserNameLinked(User user) {
        return "[" + getFoloUserName(user) + "](tg://user?id=" + user.getId() + ")";
    }

    /**
     * Получение кликабельного имени фолопидора
     *
     * @param foloPidor {@link FoloPidorDto}
     * @return Имя фолопидора
     */
    public String getFoloUserNameLinked(FoloPidorDto foloPidor) {
        return "[" + getFoloUserName(foloPidor) + "](tg://user?id=" + foloPidor.getId().getUserId() + ")";
    }

    /**
     * Проверка что {@link User} это этот бот
     *
     * @param user {@link User}
     * @return да/нет
     */
    public boolean isSelf(User user) {
        try {
            return user != null && user.getId().equals(foloBot.getMe().getId());
        } catch (TelegramApiException e) {
            printExeptionMessage(e);
            return false;
        }
    }
}

File: src/main/kotlin/com/telegram/folobot/service/UserService.kt
package com.telegram.folobot.service

import com.telegram.folobot.Utils.printExeptionMessage
import com.telegram.folobot.model.dto.FoloPidorDto
import com.telegram.folobot.model.dto.FoloUserDto
import org.springframework.stereotype.Component
import org.telegram.telegrambots.meta.api.methods.groupadministration.GetChatMember
import org.telegram.telegrambots.meta.api.objects.User
import org.telegram.telegrambots.meta.api.objects.chatmember.ChatMember
import org.telegram.telegrambots.meta.exceptions.TelegramApiException

@Component
class UserService(private val foloUserService: FoloUserService) {

    lateinit var foloBot: FoloBot

    /**
     * Получение имени пользователя
     *
     * @param user [User]
     * @return Имя пользователя
     */
    fun getFoloUserName(user: User): String {
        val foloUser: FoloUserDto = foloUserService.findById(user.id)
        // По тэгу
        var userName: String = foloUser.tag
        // Получение динамически
        if (userName.isEmpty()) userName = user.getName() ?: ""
        // По сохраненному имени
        if (userName.isEmpty()) userName = foloUser.name
        // Если не удалось определить
        if (userName.isEmpty()) userName = "Загадочный незнакомец"
        return userName
    }

    /**
     * Получение имени фолопидора
     *
     * @param foloPidorDto [FoloPidorDto]
     * @return Имя фолопидора
     */
    fun getFoloUserName(foloPidorDto: FoloPidorDto, chatId: Long): String {
        // По тэгу
        var userName: String = foloPidorDto.getTag()
        // По пользователю
        if (userName.isEmpty()) userName = getChatMember(foloPidorDto.id.userId, chatId)?.user?.getName() ?: ""
        // По сохраненному имени
        if (userName.isEmpty()) userName = foloPidorDto.getName()
        // Если не удалось определить
        if (userName.isEmpty()) userName = "Загадочный незнакомец"
        return userName
    }

    fun getChatMember(userId: Long, chatId: Long = userId): ChatMember? {
        return try {
            foloBot.execute(GetChatMember(chatId.toString(), userId))
        } catch (ignored: TelegramApiException) {
            null
        }
    }

    /**
     * Получение кликабельного имени пользователя
     *
     * @param user [User]
     * @return Имя пользователя
     */
    fun getFoloUserNameLinked(user: User): String {
        return "[" + getFoloUserName(user) + "](tg://user?id=" + user.id + ")"
    }

    /**
     * Получение кликабельного имени фолопидора
     *
     * @param foloPidor [FoloPidorDto]
     * @return Имя фолопидора
     */
    fun getFoloUserNameLinked(foloPidor: FoloPidorDto, chatId: Long): String {
        return "[" + getFoloUserName(foloPidor, chatId) + "](tg://user?id=" + foloPidor.id.userId + ")"
    }

    /**
     * Проверка, что [User] это этот бот
     *
     * @param user [User]
     * @return да/нет
     */
    fun isSelf(user: User): Boolean {
        return try {
            user.id == foloBot.me.id
        } catch (e: TelegramApiException) {
            printExeptionMessage(e)
            false
        }
    }

    /**
     * Проверка, что пользователь состоит в чате
     * @param foloPidorDto [FoloPidorDto]
     * @param chatId [Long]
     * @return [Boolean]
     */
    fun isInChat(foloPidorDto: FoloPidorDto, chatId: Long): Boolean {
        return getChatMember(foloPidorDto.id.userId, chatId)
            ?.let { !(it.status == "left" || it.status == "kicked") } ?: false
    }
}

/**
 * Получение имени пользователя
 * @return Имя пользователя
 */
fun User.getName(): String? = "${this.firstName}${this.lastName?.let { " $it" }}" ?: this.userName
